Include MultiLayer Gas Fee information for Scroll

dghelm opened the pull request:
## **Description**

### Motivation
The changes in this PR add MultiLayer Gas Fee information for Scroll. Its goal is to meet users' expectations that have used Op stack networks like Optimism and Base. In those networks, the multi-dimensional gas (L2 Fee for normal gas stuff and L1 Fee for data availablility) is presented to the user for legacy txs, so they have a clear expectation of transaction fees. Arbitrum and Linea take another approach -- fully hide the fee breakdown from users and just call it "gas," which gives an equally consistent user experience.

For Scroll, the L2 fee (which works nearly identically to Op Stack's) is also not included in `estimate_gas` calls, but users have been reporting poor experiences using MetaMask with Scroll because it's missing much of the fee paid for transactions. This PR will go a long way to helping users understand and have a consistent experience across Op and Scroll.

Note: Currently, Scroll does not support type-2 (EIP-1559) transactions, so Optimism's type-0 tx code works sufficiently for the needs of Scroll users.

### Implementation
Key changes:
- Adds Scroll and Scroll Sepolia as networks in `shared/constants/network.ts`
- Creates a new "multiLayerFee" directory to organize L2 gas functionality (and moves optimism code into a sub dir of it), and updates imports referencing these files
- Abstracts "multilayerfeenetwork" across networks, and invokes (minimally) different logic for OpStack v Scroll chains. This also lets other chains integrate logic (or, eventually I imagine, other OpStack logic as chain designs diversify), without having to touch core logic files.
- Removes several methods from `ui/selectors/selectors.js`, moving the functionality to `ui/helpers/utils/multiLayerFee/networks.js`
  - This avoids making Hook calls against the redux state for every chainId (would be costly if OpStack reaches their 1M rollup vision, also just bloats the list of selectors)
  - Allows this list of chainIds and helper methods to be called from places that don't have access to the redux state.
  - NOTE: These methods were exported but not used elsewhere in the codebase. New methods only exported if used elsewhere.

(Thanks to @pajicf for the initial work on this PR during Scroll Alpha testnet)

## **Related issues**
None identified. Tangentially related to #21139 , but doesn't fix it. (This PR adds current legacy tx support to Scroll, doesn't extend the feature to 1559 txs.)

## **Manual testing steps**

1. Add Scroll or Scroll Sepolia networks using [Chainlist](https://chainlist.org/?testnets=true&search=scroll). 
2. If needing Scroll Sepolia tokens, use [a faucet](https://docs.scroll.io/en/user-guide/faucet/), or bridge Sepolia to Scroll Sepolia [here](https://sepolia.scroll.io/bridge). (Or reply with a wallet address and I'll send some over.)
3. Start a tx. "Send" is easiest, but the bridge linked above works, or try a swap on a [testnet Uniswap showcase](https://uniswap-showcase.sepolia.scroll.xyz/).

## **Screenshots/Recordings**

NOTE: Please see this comment for [latest screenshots](https://github.com/MetaMask/metamask-extension/pull/21733#issuecomment-1937450577).

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

![image](https://github.com/MetaMask/metamask-extension/assets/4205837/828c7a30-58f4-4977-9ee4-bbc210a8c596)

Send:
![image](https://github.com/MetaMask/metamask-extension/assets/4205837/7a6fa140-60ee-498b-b3f7-9585c93a8fb2)
![image](https://github.com/MetaMask/metamask-extension/assets/4205837/4c2646b8-49e5-4842-aae0-8e8dc68b9de6) 
![image](https://github.com/MetaMask/metamask-extension/assets/4205837/ff7280ce-049d-4b8e-9aa6-1276ff79c18b)



<!-- [screenshots/recordings] -->

### **After**
![image](https://github.com/MetaMask/metamask-extension/assets/4205837/78021e2d-2a77-4db4-9438-f1bf0cf77f16)

Send:
![image](https://github.com/MetaMask/metamask-extension/assets/4205837/34b98d73-b3ee-49f1-95d3-c86605416578)
![image](https://github.com/MetaMask/metamask-extension/assets/4205837/df96d9c3-e18b-49cf-9a20-9e46c78183bc)
![image](https://github.com/MetaMask/metamask-extension/assets/4205837/16323188-c044-418f-857b-40aaa258bc6d)



<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

dghelm committed:
move multilayerfee network check logic to util methods

dghelm committed:
move optimism utils to multilayerfee and add Scroll network selector

dghelm committed:
clean console log and update test

dghelm committed:
update network svg

dghelm committed:
update tsdocs

dghelm commented:
I have read the CLA Document and I hereby sign the CLA

dghelm committed:
Merge branch 'develop' into scroll-multilayer-network

dghelm committed:
Merge branch 'develop' into scroll-multilayer-network

dghelm committed:
Merge branch 'develop' into scroll-multilayer-network

dghelm committed:
Merge branch 'develop' into scroll-multilayer-network

dghelm commented:
Hi @bschorchit -- thanks for tagging the issue. As a new contributor to MetaMask, I'm unsure of the process at this point. Is there anything I can do to help bring attention to this PR?

bschorchit commented:
Hi @dghelm, sorry for the delay in getting to this PR. We'll be prioritizing this soon (expect 1-2 weeks).

dghelm commented:
Excited to have a review, but when I went to pull in the `develop` branch, I guess there are changes that have broken my implementation -- new view:
![image](https://github.com/MetaMask/metamask-extension/assets/4205837/a6a3f4ce-3d0b-45c9-8c22-7498a815053a)

I'll try to get a fix submitted quickly.

dghelm commented:
Okay, I started to make minimal changes, but the last PR for Op Bedrock support seems to be pretty regressive for legacy txs. Here's what a legacy tx currently looks like on the `develop` branch: 
![image](https://github.com/MetaMask/metamask-extension/assets/4205837/609d40cb-9d43-4123-9ec0-427b518ba0f7)

Note: states aren't shared between "details" pane and the main pane, labels are mismatched, and different fields don't add up correctly. Also, the information is displayed twice which is odd in itself.

My suggested changes, which I'm not sure how intrusive would need to be:
- Change "Optimism Fee" label to "Execution Fee". We could say "Layer 2 Fee," but this seemed like a nice generic middle ground before gas fees get more complex to include additional dimensions. Similarly, "Layer 1 Fee" might be better labeled "L1 Data Fee".
- Change the fees above "details" to only state "estimated fee" (sum of Execution Fee and L1 Data Fee")
- Explore making "Details" component update its state as L2 gas fees or L1 data fees are updated.

dghelm committed:
Merge branch 'develop' into scroll-multilayer-network

dghelm committed:
add updates and fixes from scroll-multilayer-network-ui

dghelm committed:
remaining fixes for fixing branch divergence

dghelm committed:
fixes for Scroll Oracle needing signature and type 2 txs for Op

dghelm commented:
A number of updates made to adjust for the new UI changes.

Fixes:
- in type2 transactions, Total gas now includes L1 & L2 fees.

Changes:
- renamed "Optimism Fee" to "Execution Fee"
- dropped "multilayer fee message" and instead let the details pane show this information. Now showing a single "estimate" just like type-2 transactions

Follow-Up:
- state management is fairly messy across components for feature, and could likely be better optimized by treating it like tx gas estimates, but I imagine this is outside the scope of an outside contribution (or at least the ability of this outside contributor :) )

Example Screenshots (Before is above):

Initial confirmation window (legacy tx):
![image](https://github.com/MetaMask/metamask-extension/assets/4205837/36dccc27-986a-44a7-b9ef-9e3113d2d6d9)

Showing details:
![image](https://github.com/MetaMask/metamask-extension/assets/4205837/1c035118-dc6a-4440-9808-03b371912b35)

Showing OP type-2 tx, with correct final "Total" calculated including Layer 1 fee:
![image](https://github.com/MetaMask/metamask-extension/assets/4205837/2c6dd2ff-4d92-47bc-9e87-80b34dccdd94)

dghelm commented on ui/helpers/utils/multiLayerFee/networks.ts:
```diff
@@ -0,0 +1,53 @@
+ * @returns If chain is built on the OPStack
+ */
+export function isOpStack(chainId: string) {
+  return isOptimism(chainId) || isBase(chainId) || isOpbnb(chainId);
```
Should probably add Zora here if #20501 is getting added. Happy to pull key changes in to this PR.

bschorchit commented:
Thank you for the latest addition and contributions, the team will be wrapping up a work this week that is currently blocking moving on to this. Sorry for the delay, but we're planning to get to this soon.

icemelon commented:
@bschorchit Curious if you have any update on the blocking work?

dghelm committed:
Merge branch 'develop' into scroll-multilayer-network

dghelm commented:
New push to resolve merge conflicts (updating for Op Sepolia testnet support).

@bschorchit please let me know if there's anything else on my end to keep an eye on with Dencun upgrades :)

HowardBraham committed:
ci: fix PRs from external contributors

HowardBraham committed:
2048

HowardBraham committed:
node-browsers-medium

legobeat committed:
Merge branch 'develop' into fix-external-contribs

HowardBraham committed:
Merge branch 'fix-external-contribs' into scroll-multilayer-network

dghelm commented:
> Upon reviewing this PR, I noticed only minor differences compared to the Optimism flow. Therefore, I will be preparing the upcoming PRs for Scroll, and your initial PR will be instrumental in this process (thank you once again).
> As soon as the final PR is prepared, I will reach out to you, @dghelm, to conduct the final manual QA.

Excited to see the progress @OGPoyraz ! Would it be helpful (or speed things up) if I or someone from our team used your Optimism code changes to start PR drafts to core and extension for the adjustments needed for Scroll?

dghelm commented:
Update: addition of Scroll to the new Transaction Controller behavior is implemented in #23991
